### PR TITLE
Fix Watcher NotificationService's secure settings

### LINKED ISSUE #35378: SecureSettings in `NotificationService` are buggy
`NotificationService` implementations (eg `SlackService`, `PagerDutyService`, ..) are buggy.

They fail the communication service if the `secure_*` variant of the settings are put in the cluster state as opposed to inside the keystore, as documented https://www.elastic.co/guide/en/elastic-stack-overview/6.3/actions-pagerduty.html#configuring-pagerduty (for eg).

The normal behavior should be to simply ignore the setting when in the cluster state. In this case the service would fail to start and complain that the setting is not found in the keystore.

### PR DESCRIPTION
**Pull request by albertzaharovits**
The `NotificationService` (base class for `SlackService`, `HipchatService` ...) has both dynamic cluster settings and `SecureSetting`s and builds the clients (`Account`) that are used to comm with external services. This commit fixes an important bug about updating/reloading any of these settings (both "Secure" or not).

This use site for reloading secure settings is unlike the others because here there are also dynamic cluster settings at play - interacting to build an object. That means that, the last update to cluster settings has to pair up with the last reload of the secure settings. This commit tries to do exactly that: when calling the cluster update settings consumer it looks for the last secure settings, and rebuilds the clients. Likewise, when reloading the secure settings it looks for the last updated cluster settings, and rebuild the clients.
It's easy enough to cache the last cluster settings update but it's messy to cache the `SecureSettings`, see `NotificationService#cacheSecureSettings`. This is however the only alternative that I could come up with, without introducing any limitations (or refactoring all clients): for example we could say that dynamic cluster settings don't take effect until "reload"-ing, because there are depended `SecureSetting` that are not available at the time of the cluster update call. But this is a too big of a change... 

Another implementation detail is how to tolerate incomplete clients, because if there are two API calls required to add/remove a client (one for the cluster settings, the other for the secure settings). I this implementation, the client builder will complain if it does not find the secure settings to pair of some client cluster seettings, but will NOT complain if there are extra secure settings that cannot be paired with cluster settings. I have chosen this approach because in 6.x there could be clients requiring only cluster seettings (using deprecated ones).

Lastly, I have tried to contain the changes, and I've managed to mostly narrow them to the `NotificationService`, without touching the clients.

Closes https://github.com/elastic/elasticsearch/issues/35378

CC @romseygeek

### CODE REVIEW

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/NotificationService.java [hunk 1]
```diff
@@ -24,14 +30,34 @@
...
+            completeSettingsBuilder.setSecureSettings(cachedSecureSettings);
+        }
+        final Settings completeSettings = completeSettingsBuilder.build();
+        // create new accounts from the latest
+        final Set<String> accountNames = getAccountNames(completeSettings);
+        this.accounts = createAccounts(completeSettings, accountNames, this::createAccount);
+        this.defaultAccount = selectDefaultAccount(completeSettings, this.accounts);
     }
 
     // Used for testing only
```
Comment (spinscale) [inline x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/NotificationService.java]: can you move this ctor up again so it does not get lost?

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/NotificationService.java [hunk 2]
```diff
@@ -67,31 +108,81 @@ public Account getAccount(String name) {
...
+            final Settings accountSettings = settings.getAsSettings(getNotificationsAccountPrefix() + accountName);
+            final Account account = accountFactory.apply(accountName, accountSettings);
+            accounts.put(accountName, account);
         }
+        return Collections.unmodifiableMap(accounts);
+    }
 
-        final String defaultAccountName = settings.get("xpack.notification." + type + ".default_account");
-        A defaultAccount;
+    private @Nullable Account selectDefaultAccount(Settings settings, Map<String, Account> accounts) {
```
Comment (spinscale) [inline x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/NotificationService.java]: how about naming this `findDefaultAccount` or `findDefaultAccountOrNull`?

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/email/EmailService.java [hunk 3]
```diff
@@ -104,7 +104,11 @@
     private final CryptoService cryptoService;
 
     public EmailService(Settings settings, @Nullable CryptoService cryptoService, ClusterSettings clusterSettings) {
-        super("email", clusterSettings, EmailService.getSettings());
+        super("email", clusterSettings, Arrays.asList(SETTING_DEFAULT_ACCOUNT, SETTING_PROFILE, SETTING_EMAIL_DEFAULTS, SETTING_SMTP_AUTH,
```
Comment (spinscale) [inline x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/email/EmailService.java]: this is a lot of duplication from the setting. If a new settings gets added it needs to be added here as well. Would it make sense to have `EmailService.getSettings()` and `EmailService.getSecureSettings()` for each of the service? This also applies for the next changes

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/NotificationService.java [hunk 4]
```diff
@@ -67,31 +108,81 @@ public Account getAccount(String name) {
...
-            if (defaultAccount == null) {
+            final Account account = accounts.get(defaultAccountName);
+            if (account == null) {
                 throw new SettingsException("could not find default account [" + defaultAccountName + "]");
             }
+            return account;
+        }
+    }
+
+    private SecureSettings cacheSecureSettings(Settings source) throws GeneralSecurityException {
```
Comment (spinscale) [inline x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/NotificationService.java]: I think some more javadoc/explanation why this is done would be super useful for the next person reading this :-)

**Review by spinscale — CHANGES_REQUESTED** [review]
I left some comments, mainly cosmetic though

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/NotificationService.java (hunk 2 above)
Comment (albertzaharovits) [reply x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/NotificationService.java]: I have went with `findDefaultAccountOrNull`

**Comment by albertzaharovits** [comment]
@spinscale I have addressed all your suggestions. Mind you reappraise this, please?

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/email/EmailService.java [hunk 5]
```diff
@@ -108,11 +110,7 @@
     private final CryptoService cryptoService;
 
     public EmailService(Settings settings, @Nullable CryptoService cryptoService, ClusterSettings clusterSettings) {
-        super("email", clusterSettings, Arrays.asList(SETTING_DEFAULT_ACCOUNT, SETTING_PROFILE, SETTING_EMAIL_DEFAULTS, SETTING_SMTP_AUTH,
-                SETTING_SMTP_STARTTLS_ENABLE, SETTING_SMTP_STARTTLS_REQUIRED, SETTING_SMTP_HOST, SETTING_SMTP_PORT, SETTING_SMTP_USER,
-                SETTING_SMTP_PASSWORD, SETTING_SMTP_TIMEOUT, SETTING_SMTP_CONNECTION_TIMEOUT, SETTING_SMTP_WRITE_TIMEOUT,
-                SETTING_SMTP_SSL_TRUST_ADDRESS, SETTING_SMTP_LOCAL_ADDRESS, SETTING_SMTP_LOCAL_PORT, SETTING_SMTP_SEND_PARTIAL,
-                SETTING_SMTP_WAIT_ON_QUIT));
+        super("email", settings, clusterSettings, getClusterSettings());
```
Comment (spinscale) [inline x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/email/EmailService.java]: the naming might be a bit confusing: `clusterSettings` and then `getClusterSettings()` as variable names

**Review by spinscale — APPROVED** [review]
One last naming nit, but LGTM otherwise

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/email/EmailService.java (hunk 5 above)
Comment (albertzaharovits) [reply x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/email/EmailService.java]: I have renamed `getClusterSettings` to `getDynamicSettings` . The fact that they are cluster settings is not relevant in this context (they are all node scope) what is relevant is that they are dynamic (Property.Dynamic).

**Comment by albertzaharovits** [comment]
run gradle build test 1
and
run gradle build test 2

**Comment by albertzaharovits** [comment]
run gradle build tests 1

**Comment by albertzaharovits** [comment]
test this please

**Comment by albertzaharovits** [comment]
run gradle build tests 1

**Comment by albertzaharovits** [comment]
run gradle build tests 2

**Pull request merged**